src/relax/transform/dataflow_inplace.cc

```diff
@@ -0,0 +1,1038 @@
+    // now because later ops might need the same one. Instead, we will clean up at the end
+    auto new_gv = builder_->AddFunction(legal_primfunc, inline_legal_op_name);
+
+    // update the call (change the op, update the argument, change the attrs)
+    legalized_call_cow->op = call_tir_inplace_op;
+
+    Array<Expr> new_args(legalized_call->args.begin(), legalized_call->args.end());
+    new_args.Set(0, new_gv);
+    legalized_call_cow->args = new_args;
+
+    ObjectPtr<CallTIRInplaceAttrs> attrs = make_object<CallTIRInplaceAttrs>();
+    attrs->inplace_indices = inplace_indices;
+    legalized_call_cow->attrs = Attrs(attrs);
+
+    return legalized_call;
+  }
+
+  // Made public for testing.
+  IRModule CurrentMod() { return builder_->GetContextIRModule(); }
+
+ private:
+  const IRModule& mod_;
+  Array<GlobalVar>
+      legalizers_added;    // Keep track of legalizers we add so we can clean up at the end.
+  Array<Var> func_params;  // The current function's params will be treated as non-aliased
+                           // (we are assuming good behavior on the user's part).
+};
+
+namespace transform {
+
+Map<Var, Array<Integer>> DataflowLivenessAnalysis(const DataflowBlock& block) {
+  auto liveness_ranges = AnalyzeLiveness(block);
+  Map<Var, Array<Integer>> ret;
+  for (auto kv : liveness_ranges) {
+    ret.Set(kv.first, {kv.second.first, kv.second.second});
+  }
+  return ret;
+}
+
+Array<ObjectRef> DataflowAliasAnalysis(const DataflowBlock& block, Array<Var> inputs) {
+  AliasAnalyzer analyzer;
+  auto res = analyzer.Analyze(block, inputs);
+  auto alias_sets = res.first;
+  auto tuple_map = res.second;
+  Map<Var, Array<Integer>> new_alias_sets;
+  Map<Integer, Array<Array<Integer>>> new_tuple_map;
+  for (auto kv : alias_sets) {
+    Array<Integer> aliases;
+    for (auto alias : kv.second) {
+      aliases.push_back(alias);
+    }
+    new_alias_sets.Set(kv.first, aliases);
+  }
+  for (auto kv : tuple_map) {
+    Array<Array<Integer>> elem_aliases;
+    for (auto alias_set : kv.second) {
+      Array<Integer> dim_aliases;
+      for (auto alias : alias_set) {
+        dim_aliases.push_back(alias);
+      }
+      elem_aliases.push_back(dim_aliases);
+    }
+    new_tuple_map.Set(kv.first, elem_aliases);
+  }
+  return {new_alias_sets, new_tuple_map};
+}
+
+// this would be preferable to do as a dataflow block pass,
+// but the transformation adds new PrimFuncs, so it affects the module
+tvm::transform::Pass DataflowUseInplaceCalls() {
+  auto inplace_pass = tvm::transform::CreateModulePass(
+      [](const IRModule& mod, const PassContext& ctx) -> IRModule {
+        ModuleInplaceTransformer transformer(mod);
+        return transformer.Transform();
+      },
+      0, "DataflowInsertInPlaceCalls", {}, false);
+  // odd quirk: if Normalize is not explicitly called, then the function
+  // StructInfo will not be properly updated
+  return tvm::transform::Sequential({inplace_pass, Normalize()});
```

Comment:
Not sure why this happened. I assumed it was because the `BlockBuilder` was not renormalizing when doing an update, but called `BlockBuilder::Normalize()` before updating the function also did not fix this issue.

Reply:
Maybe this can be avoided if I use `BlockBuilder` methods throughout the transformation. I will see if I can write it that way.

Reply:
As another option, it may be easier to rely on the `BlockBuilder` usage in the base `ExprMutator` class. Looks like there are two parts that would need to be updated.

1. Normalization of bindings within `DataflowBlock`
    a. The `VisitBindingBlock_(const DataflowBlockNode* op)` would need perform the analysis, save the result in a member variable, then delegate to `ExprMutator::VisitBindingBlock_`.
    b. The updates would need to occur in `Expr VisitExpr_(const CallNode* op)`.  It would check whether there is a replacement planned in the member variable from (a), returning either the updated object or the original.  Normalization would occur within the base class `ExprMutator::VisitBinding_(const VarBinding*, const CallNode*)`.
    
2. Normalization of `relax::Function`.

   ```c++
    // Instead of
    auto function = GetRef<Function>(func_node);                                       
    auto* function_cow = function.CopyOnWrite();                                       
    auto new_body = VisitExpr(function->body);                                         
                       
    // Replace with                                                               
    auto function = Downcast<Function>(VisitExpr(GetRef<Function>(func_node)));
```
     The copy-on-write can be handled internally by `ExprMutator`, if any mutations occur.

Reply:
I tried using `BlockBuilder` methods to build up the new dataflow block and still had an error if I didn't have the explicit normalization in here. I'm not sure what exactly is going wrong. Specifically, the error had to do with the `StructInfo` of the `SeqExpr`'s body, but calling normalize on the function's body only did not solve the problem. 

@yongwww, you mentioned that you've encountered an issue like this before. Do you know what could be going wrong here?

Reply:
Thank you, @Lunderberg, I had tried step 1 but not also step 2. Using the function node visitor fixed everything. There are lots of invariants in `ExprMutator` to know about!

Reply:
No problem!  I think it's a bit tricky with how the `relax::ExprMutator` exposes a similar interface to the `tir::ExprMutator`, but can have extra requirements in some cases.  I think it's helpful to have the automatic normalization and struct info inference, but it does seem a bit more fragile.

My personal rule of thumb is to implement everything in terms of `VisitVarDef` and `VisitExpr_` overrides whenever possible, as those don't include any interactions with the `block_builder_` in the base `ExprMutator` implementation.